ddtrace/internal/module.py

```diff
@@ -0,0 +1,427 @@
+        # spec instead
+        try:
+            orig = abspath(module.__spec__.origin)  # type: ignore
+        except (AttributeError, ValueError, TypeError):
+            orig = None
+
+    if orig is not None and isfile(orig):
+        if orig.endswith(".pyc"):
+            orig = orig[:-1]
+        return orig
+
+    return "<unknown origin>"
+
+
+def _resolve(path):
+    # type: (str) -> Optional[str]
+    """Resolve a (relative) path with respect to sys.path."""
+    for base in sys.path:
+        if isdir(base):
+            resolved_path = abspath(join(base, expanduser(path)))
+            if isfile(resolved_path):
+                return resolved_path
+    return None
+
+
+# Borrowed from the wrapt module
+# https://github.com/GrahamDumpleton/wrapt/blob/0baff1b6ac8d64ffb08ea2d1610c7ed90115b9d5/src/wrapt/importer.py
+class _ImportHookChainedLoader:
+    def __init__(self, loader, callback):
+        # type: (Loader, Callable[[ModuleType], None]) -> None
+        self.loader = loader
+        self.callback = callback
+
+    def load_module(self, fullname):
+        # type: (str) -> ModuleType
+        module = self.loader.load_module(fullname)
+        self.callback(module)
+
+        return module
+
+    def get_code(self, mod_name):
+        return self.loader.get_code(mod_name)
+
+
+class ModuleWatchdog(dict):
+    """Module watchdog.
+
+    Replace the standard ``sys.modules`` dictionary to detect when modules are
+    loaded/unloaded. This is also responsible for triggering any registered
+    import hooks.
+    """
+
+    _run_code = None
+
+    def __init__(self):
+        # type: () -> None
+        self._hook_map = defaultdict(list)  # type: DefaultDict[str, List[Tuple[HookType, Any]]]
+        self._origin_map = {origin(module): module for module in sys.modules.values()}
+        self._modules = sys.modules
+        self._finding = set()  # type: Set[str]
+
+    def __enter__(self):
+        # type: () -> ModuleWatchdog
+        self._add_to_meta_path()
+        sys.modules = self
+        log.debug("%s installed", type(self))
+        return self
+
+    def __exit__(self, *exc):
+        # type: (Tuple[Any]) -> None
+        self._remove_from_meta_path()
+        sys.modules = self._modules
+        log.debug("%s uninstalled", type(self))
+
+    def __getitem__(self, item):
+        # type: (str) -> ModuleType
+        return self._modules.__getitem__(item)
+
+    def __setitem__(self, name, module):
+        # type: (str, ModuleType) -> None
+        self._modules.__setitem__(name, module)
+
+    def _add_to_meta_path(self):
+        # type: () -> None
+        sys.meta_path.insert(0, self)
+
+    def _remove_from_meta_path(self):
+        # type: () -> None
+        for i, meta_path in enumerate(sys.meta_path):
+            if type(meta_path) is type(self):
+                sys.meta_path.pop(i)
+                return
+
+    def after_import(self, module):
```

Comment:
```suggestion
    def _after_import(self, module):
```

shouldn't be in the public API?

Reply:
In fact, the idea is that this method should be overwritten by subclasses to implement custom behaviour. That's why I didn't add the `_`.